Marlin/planner.cpp

```diff
@@ -2521,23 +2651,41 @@ bool Planner::buffer_segment(const float &a, const float &b, const float &c, con
  * On CORE machines stepper ABC will be translated from the given XYZ.
  */
 
-void Planner::_set_position_mm(const float &a, const float &b, const float &c, const float &e) {
+void Planner::_set_position_mm(const float &a, const float &b, const float &c
+  #if ENABLED(HANGPRINTER)
+    , const float &d
+  #endif
+  , const float &e
+) {
   #if ENABLED(DISTINCT_E_FACTORS)
     last_extruder = active_extruder;
   #endif
-  position[A_AXIS] = LROUND(a * axis_steps_per_mm[A_AXIS]);
-  position[B_AXIS] = LROUND(b * axis_steps_per_mm[B_AXIS]);
-  position[C_AXIS] = LROUND(axis_steps_per_mm[C_AXIS] * (c + (
-    #if !IS_KINEMATIC && ENABLED(AUTO_BED_LEVELING_UBL)
-      leveling_active ? ubl.get_z_correction(a, b) :
+  #if ENABLED(LINE_BUILDUP_COMPENSATION_FEATURE)
+    position[A_AXIS] = LROUND(k0[A_AXIS] * (SQRT(k1[A_AXIS] + a * k2[A_AXIS]) - sqrtk1[A_AXIS])),
+    position[B_AXIS] = LROUND(k0[B_AXIS] * (SQRT(k1[B_AXIS] + b * k2[B_AXIS]) - sqrtk1[B_AXIS])),
+    position[C_AXIS] = LROUND(k0[C_AXIS] * (SQRT(k1[C_AXIS] + c * k2[C_AXIS]) - sqrtk1[C_AXIS])),
+    position[D_AXIS] = LROUND(k0[D_AXIS] * (SQRT(k1[D_AXIS] + d * k2[D_AXIS]) - sqrtk1[D_AXIS])),
+  #else
+    position[A_AXIS] = LROUND(a * axis_steps_per_mm[A_AXIS]);
+    position[B_AXIS] = LROUND(b * axis_steps_per_mm[B_AXIS]);
+    position[C_AXIS] = LROUND(axis_steps_per_mm[C_AXIS] * (c + (
+      #if !IS_KINEMATIC && ENABLED(AUTO_BED_LEVELING_UBL)
+        leveling_active ? ubl.get_z_correction(a, b) :
+      #endif
+      0)
+    ));
```

Comment:
@tobbelobb — This is one of the new changes I wasn't sure how to adapt to `HANGPRINTER`. UBL leveling is applied to the C axis when directly setting the position in the planner. This is only needed for UBL, and not the other leveling options, because it is unique in not being tightly-integrated in the planner.